test/jdk/jdk/internal/platform/docker/TestDockerMemoryMetrics.java

```diff
@@ -110,7 +110,14 @@ private static void testMemoryFailCount(String value) throws Exception {
                 .addJavaOpts("-cp", "/test-classes/")
                 .addJavaOpts("--add-exports", "java.base/jdk.internal.platform=ALL-UNNAMED")
                 .addClassOptions("failcount");
-        DockerTestUtils.dockerRunJava(opts).shouldHaveExitValue(0).shouldContain("TEST PASSED!!!");
+
+        OutputAnalyzer oa = DockerTestUtils.dockerRunJava(opts);
+        String output = oa.getOutput();
+        if (output.contains("Not OOM killed")) {
+            if (!output.contains("Ignoring test")) {
+                oa.shouldHaveExitValue(0).shouldContain("TEST PASSED!!!");
+            }
+        }
```

Comment:
Consider a broken implementation of `Metrics.systemMetrics().getMemoryFailCount()`. Wouldn't the test now (falsely) pass?

What is the actual test output on those systems where the test fails? There should be a docker log file. Does it enter line 91?

Reply:
Thanks @jerboaa for your review.

Test output when failing is just:
```
[failcount] 
```

It doesn't enter line 91 of test/jdk/jdk/internal/platform/docker/MetricsMemoryTester.java.


I assume memory.failcnt has been disabled on these platforms since it's always 0.
What about checking the value of memory.failcnt after testMemoryFailCount like:
```
// check after testMemoryFailCount()
if (memory.failcnt is zero) {
  // memory.failcnt has been disabled
  pass
} else {
  // a broken implementation of Metrics.systemMetrics().getMemoryFailCount()
  fail
}
```
Thanks.

Reply:
Hi @jerboaa ,

After more thinking, I think the reason why memory.failcnt is always 0 is that there is no swap space on the host machine.
So the testMemoryFailCount should be skipped in that case.

But is there any API which can be used to get the swap space size of the host machine?

Thanks.

Reply:
> Consider a broken implementation of `Metrics.systemMetrics().getMemoryFailCount()`. Wouldn't the test now (falsely) pass?

Hi @jerboaa ,

A pre-test run has been added to check whether swapping really works for testMemoryFailCount.
Swapping should be OK for memory.failcnt testing otherwise it will fail due to OOM killed.

What do you think?
Thanks.

Reply:
@DamonFool Hmm, if swap not working is the issue the test shouldn't enter the branch which you say is failing. From `MetricsMemoryTester.java`:

```
    private static void testMemoryFailCount() {
        long memAndSwapLimit = Metrics.systemMetrics().getMemoryAndSwapLimit();
        long memLimit = Metrics.systemMetrics().getMemoryLimit();

        // We need swap to execute this test or will SEGV
        if (memAndSwapLimit <= memLimit) { // <=============== This is checking whether or not swap works
            System.out.println("No swap memory limits, test case skipped");
        } else {
```

It has been added with JDK-8250984. So now I'm even more confused what's going on here...

Reply:
> if (memAndSwapLimit <= memLimit)

Hi @jerboaa ,

Unfortunately, this check fails to work on a host machine which doesn't support swapping (e.g., total size of Swap is 0 byte).

You can still specify --memory and --memory-swap as you like on a host machine without swapping space.
And Metrics.systemMetrics().getMemoryLimit()/getMemoryAndSwapLimit() do return the values as you specified.
But the swapping will never happen since there is no swapping space on the host machine.

Thanks.